### ISSUE TITLE
Support PostCSS

### ISSUE THREAD

**Issue by nnecec** [huozhi/bunchee#440]
Bunchee supports CSS now, but I think we need to at least support Tailwind. So supporting PostCSS is a better idea.

## TL;DR

Create an internal postcss-plugin. It executes the postcss API to transform CSS code.

## Features

1. Detect postcss is installed, then try to execute postcss API to transform CSS code.
2. Injecting CSS code into a js file is the default option now, we could create an option --extract-css to support generating a separate CSS file. (or generate separate css file by default)

## How to implement it?

Create an internal postcss-plugin that will check if postcss is installed and if there is a postcss.config configuration. If not, it will return the code directly. If so, the plugin reads poctcss.config through postcss-config-load and passes options and plugins to the PostCSS API for transformation. About the transformed code, we can use inline-css's logic to inject it into js files.

### Why don't use [rollup-plugin-postcss](https://www.npmjs.com/package/rollup-plugin-postcss)?

rollup-plugin-postcss is no longer maintained for over 2 years. It provides sass-loader and less-loader which I don't think are necessary.

## Example

Here is an example of the input and output.

```bash
bunchee --extract-css

# files structure

package
├── dist
│   ├── index.css
│   ├── index.js
│   └── index.js.map
├── src
│   ├── index.css
│   └── index.ts
├── package.json
├── postcss.config.js
├── tailwind.config.js
└── tsconfig.json
```

**Comment by nnecec** [comment]
rollup-plugin-postcss is no longer maintained.It provide sass-loader and less-loader, I think we don't need these loader.

**Comment by nnecec** [comment]
TL;DR

Create an internal postcss-plugin. It executes the postcss API to transform CSS code. 

Why don't use [rollup-plugin-postcss](https://www.npmjs.com/package/rollup-plugin-postcss)?

`rollup-plugin-postcss` is no longer maintained for over 2 years. It provides sass-loader and less-loader which I don't think are necessary.

How to implement it?

Create an internal postcss-plugin that will check if `postcss` is installed and if there is a `postcss.config` configuration. If not, it will return the code directly. If so, the plugin reads `poctcss.config` through `postcss-config-load` and passes options and plugins to the PostCSS API for transformation. About the transformed code, we can use inline-css's logic to inject it into js files.

**Comment by huozhi** [comment]
Can you share some examples of code with tailwind as input, then what's the expected in the output bundle?

plus: let's group everything of RFC in the issue description instead of comments, to help others read easier

**Comment by nnecec** [comment]
I updated the description of the issue and added an example of input and output.

**Comment by huozhi** [comment]
The problem is how the css is going to be declared in package.json. `"style"` property is not a standard thing, not all the bundler will pick it up.
Other questions:
If there're multiple exports, and each one uses different css files, how we gonna output them? 
If there's an barrel file, an index.js that included all the exports, does it mean we gonna bundle all the CSS together?

**Comment by emirefek** [comment]
Hi these issue seems like forgotten but I think it is a very needed feature.

> The problem is how the css is going to be declared in package.json. `"style"` property is not a standard thing, not all the bundler will pick it up.
Yeap this is not a standart thing but most bundlers are figure it out unstandart way. Here is a example I am using to build tailwind css modules with bunchee.

```json
// file: package.json
{
  "name": "@repo/ui",
  "files": [
    "dist/**",
    "dist"
  ],
  "exports": {
    "./button": {
      "import": {
        "types": "./dist/es/button.d.mts",
        "default": "./dist/es/button.mjs"
      },
      "require": {
        "types": "./dist/cjs/button.d.ts",
        "default": "./dist/cjs/button.js"
      }
    },
    "./styles.css": {
      "import": "./dist/styles.css",
      "require": "./dist/styles.css"
    }
  },
  "scripts": {
    "build": "tsx ./builder.ts",
    "dev": "tsx ./builder.ts --watch",
    "build:tw": "pnpx @tailwindcss/cli -i ./src/styles.css -o ./dist/styles.css",
    "dev:tw": "pnpx @tailwindcss/cli -i ./src/styles.css -o ./dist/styles.css --watch",
    "dev:js": "bunchee --watch",
  },
  ...
}
```

```ts
// file: builder.ts

import { BundleConfig, bundle } from "bunchee";
import { readFile, writeFile } from "fs/promises";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import { execSync, spawn } from "child_process";

const bundleConfig: BundleConfig = {};

async function main() {
  const isWatchMode = process.argv.includes("--watch");

  if (isWatchMode) {
    await watchFlow();
  } else {
    await buildFlow();
  }
}

async function buildFlow() {
  console.log("Starting build mode...");

  const packageJsonPath = resolve(
    dirname(fileURLToPath(import.meta.url)),
    "./package.json",
  );

  // read package.json
  const packageJson = JSON.parse(await readFile(packageJsonPath, "utf-8"));

  // remove the './styles.css' key inside 'exports' object of package.json
  const stylesBackup = packageJson.exports["./styles.css"];
  delete packageJson.exports["./styles.css"];

  // write it back to package.json
  await writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2));

  // Give the file system time to complete the write operation
  await new Promise((resolve) => setTimeout(resolve, 100));

  // Execute commands with stdio inheritance to display output
  execSync(`pnpx bunchee`, { stdio: "inherit" });

  // now we need to restore the package.json
  packageJson.exports["./styles.css"] = stylesBackup;
  await writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2));

  // Give the file system time to complete the write operation
  await new Promise((resolve) => setTimeout(resolve, 100));

  execSync("pnpx @tailwindcss/cli -i ./src/styles.css -o ./dist/styles.css", {
    stdio: "inherit",
  });

  console.log("Build complete!");
}

async function watchFlow() {
  console.log("Starting watch mode...");
  // Use similar logic to buildFlow but with watch flags

  const packageJsonPath = resolve(
    dirname(fileURLToPath(import.meta.url)),
    "./package.json",
  );
  const currentDir = dirname(fileURLToPath(import.meta.url));
  const packageJson = JSON.parse(await readFile(packageJsonPath, "utf-8"));
  const stylesBackup = packageJson.exports["./styles.css"];
  delete packageJson.exports["./styles.css"];
  await writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2));

  // Try using execSync to start the bunchee process in watch mode in the background
  try {
    console.log("Starting bunchee in watch mode...");
    const buncheeProcess = spawn("pnpx", ["bunchee", "--watch"], {
      stdio: "inherit",
      cwd: currentDir,
      shell: true,
      detached: false,
    });

    // Add error handling for the process
    buncheeProcess.on("error", (err) => {
      console.error("Failed to start bunchee process:", err);
    });

    console.log("Starting tailwind in watch mode...");
    const tailwindProcess = spawn(
      "pnpx",
      [
        "@tailwindcss/cli",
        "-i",
        "./src/styles.css",
        "-o",
        "./dist/styles.css",
        "--watch",
      ],
      {
        stdio: "inherit",
        cwd: currentDir,
        shell: true,
      },
    );

    tailwindProcess.on("error", (err) => {
      console.error("Failed to start tailwind process:", err);
    });

    console.log("Watch mode active. Waiting for changes...");

    // Give the file system time to complete the write operation
    await new Promise((resolve) => setTimeout(resolve, 2000));

    // Restore package.json after bunchee has started
    packageJson.exports["./styles.css"] = stylesBackup;
    await writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2));

    // Handle process termination gracefully
    process.on("SIGINT", () => {
      console.log("Terminating watch processes...");
      buncheeProcess.kill();
      tailwindProcess.kill();
      process.exit();
    });
  } catch (error) {
    console.error("Error in watch mode:", error);

    // Make sure to restore package.json even if there's an error
    packageJson.exports["./styles.css"] = stylesBackup;
    await writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2));

    process.exit(1);
  }
}

main();
```

So this is the workaround I am using to export styles.css which generated by tailwind css. But it's very problematic I sometimes need to restart dev mode because things get override each other. This way allows you to use exported css in both "nextjs" and "vite" bundler. So I think that covers most of the stuff bunchee project aims.

I don't know bunchee's code base much but if any maintainer is approach the feature request positivly I am ready to contribute.

**Comment by huozhi** [comment]
Closing this issue for now, I explained the reason in [this comment](https://github.com/huozhi/bunchee/issues/639#issuecomment-2994379902) where you can just uses tailwind v4 and no need to have extra setup for integrate your library or generate the extra CSS file.
